api.js

```diff
@@ -224,3 +236,87 @@ Api.prototype._run = function (files, _options) {
 		return runStatus;
 	});
 };
+
+var getBlankResults = function () {
+	return {
+		stats: {
+			testCount: 0,
+			passCount: 0,
+			skipCount: 0,
+			todoCount: 0,
+			failCount: 0
+		},
+		tests: []
+	};
+};
+
+Api.prototype._runLimitedPool = function (files, runStatus, concurrency) {
+	var self = this;
+	var tests = {};
+
+	runStatus.on('timeout', function () {
+		Object.keys(tests).forEach(function (file) {
+			var fork = tests[file];
+			fork.exit();
+		});
+	});
+
+	return Promise.map(files, function (file) {
+		try {
+			var test = tests[file] = self._runFile(file, runStatus);
+
+			return new Promise(function (resolve) {
+				var runner = self._generateRunner(file, test, runStatus, resolve);
```

Comment:
Currently `_generateRunner` does a bit too much. I think once it's simplified it'll just fit here instead.